### ISSUE TITLE
outbound-networking does not work as expected for MySQL

### ISSUE THREAD

**Issue by ThorstenHans** [fermyon/spin#2113]
## Context

I created a Spin application that interacts with an MySQL database. In my case, the database is hosted in Azure (but this does not matter at all). I recognized outbound connectivity being blocked when the MySQL password contains a `#` (hash sign)! 

In my case, `allowed_outbound_hosts` in `Spin.toml` look like this: 

```toml
allowed_outbound_hosts = ["mysql://mydatabase.mysql.database.azure.com"]
``` 

I use the `Connection` struct from `spin_sdk::mysql` to establish the connection as shown below:

```rust
let c = Connection::open("mysql://thorsten:pass#word001@mydatabase.mysql.database.azure.com/testdb?ssl-mode")?;
```

At runtime, I get the following error:

```bash
Handler returned an error: Error::ConnectionFailed("address mysql://thorsten:pass#word001@mydatabase.mysql.database.azure.com/testdb?ssl-mode=required is not permitted")
```

On top of permitting the outgoing connection, my password is not shown in `stdout`. 

## Root Cause

I checked the source to spot the root cause, which is ultimately here:

https://github.com/fermyon/spin/blob/9e7a68d17f82b51239c4217b57948d8df98db3a0/crates/outbound-networking/src/lib.rs#L302C9-L316C12

My guess is that the `url` crate, considers the `#` from the password introducing a URL hash-fragment. That being said, I think this may also break for outgoing connections with credentials in the connection string and other schemes than `mysql`. 

The behavior could be reproduced by adding the following unit test:

```rust
#[test]
    fn test_hash_char_in_db_password() {
        let allowed = AllowedHostsConfig::parse(&["mysql://xyz.com"]).unwrap();
        assert!(
            allowed.allows(&OutboundUrl::parse("mysql://user:pass#word@xyz.com", "mysql").unwrap())
        );
    }
```

## Environment

os: `Darwin mbp14.local 23.1.0 Darwin Kernel Version 23.1.0: Mon Oct  9 21:27:24 PDT 2023; root:xnu-10002.41.9~6/RELEASE_ARM64_T6000 arm6`
spin: `spin 2.0.1 (1d72f1c 2023-11-10)`
spin-sdk: `{ git = "https://github.com/fermyon/spin", tag = "v2.0.0" }`

**Comment by ThorstenHans** [comment]
Adding @rylev to this one.

I thought about checking if the scheme is not `http` and not `https`, if not, I would remove username and password (if present) from the string before parsing it using the `url` crate.

However, I thought this may be a smell. Maybe you've a more idiomatic fix in mind to address this?

**Comment by rylev** [comment]
Thanks for reporting! This is a subtly tricky one. One could make the argument (annoyingly so) that the url as currently written is incorrect since in order for the url to be properly formatted the username and password must be url encoded. 

I'm not really sure how to fix this. We could write some hacky code that tries to look for anything between `://` and `@` and sees whether it is url encoded or not - potentially warning if it's not. However, I'd be worried this might break in other ways...

cc @itowlson who might have thoughts

**Comment by ThorstenHans** [comment]
@rylev yes had the same thoughts. However, I checked with other tools like mysql-client CLI. Those work without URL-encoding the same connection string.

I spiked a solution, by matching the scheme argument of the parse function. As you also mentioned, it felt hacky. and that's why I went with the issue in the first place instead of submitting a PR 😃 

IMO, removing everything behind the scheme including the `@` seems like the most robust way for - just - parsing the URL using the `url` crate.
